### ISSUE TITLE
Rotation trick diverges when stochastic_sample_codes=True

### ISSUE THREAD

**Issue by mahdip72** [lucidrains/vector-quantize-pytorch#223]
@lucidrains Thanks for the brilliant repo. It’s been central to our new project: [link](https://github.com/mahdip72/vq_encoder_decoder) 

Recently, we are hitting divergence during training whenever `rotation_trick` and `stochastic_sample_codes=True` are both enabled, even with `sample_codebook_temp=0.01`, it does not fully solve the problem. However, disabling the `rotation_trick` restores stability. Is this combination expected to misbehave, or is there a workaround you’d recommend for us?

We are using these parameters for your reference:

```yaml
dim: 64
decay: 0.99
codebook_size: 8196
commitment_weight: 0.25
orthogonal_reg_weight: 10
orthogonal_reg_max_codes: 512
orthogonal_reg_active_codes_only: True
rotation_trick: True
threshold_ema_dead_code: 3
kmeans_init: True
kmeans_iters: 10
stochastic_sample_codes: True
sample_codebook_temp: 0.1
```

**Comment by mahdip72** [comment]
@lucidrains I tried keeping both on, but in practice, we’re seeing the same pattern across two separate codebases (GCP-VQVAE and another VQ-VAE variant): with rotation_trick=True and stochastic sampling enabled, training runs fine for a few epochs and then blows up. Dropping sample_codebook_temp to 0.01 only delays the blow-up. I asked GPT-5 Codex for a sanity check and it suggested the gradient might be getting rotated toward essentially random directions when the sampled code jumps far from the input:

> With stochastic sampling, the forward codeword can be far from the input vector. The rotation trick then forces the gradient to align with a largely random direction, amplifying updates instead of dampening them; this is consistent with the divergence you observe. When rotation_trick=False, the code falls back to STE (quantize = x + (quantize - x).detach()), so gradients stay identity-mapped and training remains stable. There is no guard to prevent the problematic combination.

Does that line up with how you expect the rotation trick to behave, or is there something else we should be looking at?

**Comment by mahdip72** [comment]
Here are the validation results I have got for different combinations in my new research project, which might be interesting for others:

**Dark Blue** plot:
```yaml
rotation_trick: False
stochastic_sample_codes: False
sample_codebook_temp: 0.01
```

**Light Blue** plot:
```yaml
rotation_trick: False
stochastic_sample_codes: True
sample_codebook_temp: 0.1
```

**Green** plot:
```yaml
rotation_trick: False
stochastic_sample_codes: True
sample_codebook_temp: 0.5
```

**Orange** plot:
```yaml
rotation_trick: False
stochastic_sample_codes: False
```

**Red** plot:
```yaml
rotation_trick: True
stochastic_sample_codes: False
```

<img width="809" height="675" alt="Image" src="https://github.com/user-attachments/assets/52cbda89-2d10-4e72-929d-b61741091d9a" />

Other hyperparameters were identical in all trainings. `sample_codebook_temp` was constant throughout the training for those who had `stochastic_sample_codes` enabled. These are my other VQ layer parameters:

```yaml
dim: 64
decay: 0.99
codebook_size: 8196
commitment_weight: 0.25
orthogonal_reg_weight: 10
orthogonal_reg_max_codes: 1024
orthogonal_reg_active_codes_only: True
threshold_ema_dead_code: 3
kmeans_init: True
kmeans_iters: 10
```

I also observed that in the rotation trick enabled (**Blue** plot below), the gradient norm of VQ loss is higher:

<img width="924" height="706" alt="Image" src="https://github.com/user-attachments/assets/b54efeb2-204a-4c87-b7a0-e7854b3ddb16" />

@lucidrains Same as before, I could not make the combination of  `rotation_trick` and `stochastic_sample_codes` work in any temperature; all blew up after some epochs.

As a conclusion, for this specific project, STD + sampling with temperature 0.1 acts better than enabling `rotation_trick`.


**Note:**
The essence of this project was about using VQVAE for quantizing the continuous representation of a pretrained foundational model.

**Comment by mahdip72** [comment]
I did the following two experiments in this setup:

```yaml
dim: 64
decay: 0.99
codebook_size: 8196
commitment_weight: 0.25
orthogonal_reg_weight: 10
orthogonal_reg_max_codes: 1024
orthogonal_reg_active_codes_only: True
rotation_trick: True
threshold_ema_dead_code: 3
kmeans_init: True
kmeans_iters: 10
stochastic_sample_codes: True
sample_codebook_temp: 0.1
```

The only difference I made in them is enabling/disabling an experimental feature named `adaptive_loss_coefficients` to mitigate divergence, which is explained [here](https://github.com/mahdip72/vq_encoder_decoder?tab=readme-ov-file#experimental-features).

**Validation set**
Red: `adaptive_loss_coefficients` was enabled.
Gray: `adaptive_loss_coefficients` was disabled.

<img width="500" height="400" alt="Image" src="https://github.com/user-attachments/assets/e0b14f63-c227-4625-a8ea-9b15a2f6b98c" />

**Training set**
Green `adaptive_loss_coefficients` was enabled.
Blue `adaptive_loss_coefficients` was disabled.


<img width="1200" height="700" alt="Image" src="https://github.com/user-attachments/assets/1f281da8-24dd-4d07-95ae-be4e724c3163" />

The reconstruction loss is computed as the sum of `cosine` and `mse` terms.
The `unscaled` term means that any scaling coefficients were removed during logging, allowing the visualization of raw, unweighted values.

<img width="1000" height="800" alt="Image" src="https://github.com/user-attachments/assets/18d7cb84-6d85-4c8a-9140-09f8975d716d" />

<img width="1000" height="800" alt="Image" src="https://github.com/user-attachments/assets/787a8f81-ef6f-45b8-90d5-454c3cbc4ff0" />

Higher `sample_codebook_temp` values only make it worse.

**Comment by Nemo-1024** [comment]
@mahdip72  Thank you for sharing the experimental results. I’m curious whether the default-enabled EMA updates might conflict with the orthogonal loss, but I don’t have experience in this area.

**Comment by Nemo-1024** [comment]
I am attempting to integrate the VQ module of this project to train a latent action model in the latent space of DINOv3. This model treats the transformations between consecutive frames as pseudo-action labels. However, I have not noticed any configuration that would automatically disable EMA when the orthogonality loss is enabled. Are you suggesting that the gradient flow from the orthogonality loss updates the codebook directly, thereby making EMA ineffective?

Additionally, I found that my codebook collapses severely (even after I tried to avoid this issue). Do you have any suggestions for training a small codebook—for example, with only 16 or 64 entries?
